### PR TITLE
check driver verifier status before enable it

### PR DESCRIPTION
**Pull request by suqinhuang**
Add function to check verifier status

### CODE REVIEW

**Comment by suqinhuang** [comment]
ID: 1336311

File: virttest/utils_test/qemu.py [hunk 1]
```diff
@@ -132,6 +132,22 @@ def pin_vm_threads(vm, node):
         logging.info("Skip pinning, no enough nodes")
 
 
+def check_driver_verifier(session, driver, vm, timeout=300):
+    """
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: why give VM param, it looks not used anywhere?

File: virttest/utils_test/qemu.py [hunk 2]
```diff
@@ -142,20 +158,25 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
     :param timeout: Timeout in seconds.
     :return: Session after reboot.
     """
-    verifier_setup_cmd = "verifier.exe /standard /driver %s.sys" % driver
-    if driver:
+    verifier_setup_cmd = "verifier /standard /driver %s.sys" % driver
+
+    (verifier_status, output) = check_driver_verifier(session,
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: useless '()'

File: virttest/utils_test/qemu.py [hunk 3]
```diff
@@ -142,20 +158,25 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
...
-
-    verifier_query_cmd = "verifier.exe /querysettings"
-    output = session.cmd_output(verifier_query_cmd, timeout=timeout)
-    if driver not in output:
-        msg = "Verify device driver failed, "
-        msg += "guest report driver is %s, " % output
-        msg += "expect is '%s'" % driver
-        raise exceptions.TestFail(msg)
+        session = vm.reboot(session)
+        (verifier_status, output) = check_driver_verifier(session,
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: same as line#163

**Comment by CongLi** [comment]
I think it should be:
1. setup verifier.exe 2. check status
No need to check status before setup verifier.exe, as it does not matter for the cmd execution.

**Comment by suqinhuang** [comment]
I think it's better to skip it if the step is useless.

Suqin

File: virttest/utils_test/qemu.py [hunk 4]
```diff
@@ -142,20 +158,25 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
     :param timeout: Timeout in seconds.
     :return: Session after reboot.
     """
-    verifier_setup_cmd = "verifier.exe /standard /driver %s.sys" % driver
-    if driver:
+    verifier_setup_cmd = "verifier /standard /driver %s.sys" % driver
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: move this line  into if branch, because it only used in if branch.

Comment (xutian) [inline virttest/utils_test/qemu.py]: other parts look ok to me. @CongLi please help to review this path.

Thanks,
Xu

File: virttest/utils_test/qemu.py [hunk 5]
```diff
@@ -167,7 +192,8 @@ def clear_win_driver_verifier(session, vm, timeout=300):
     :param timeout: Timeout in seconds.
     :return: Session after reboot.
     """
-    verifier_clear_cmd = "verifier.exe /reset"
+    error_context.context("Clear %s driver verifier" % driver)
```
Comment (CongLi) [inline virttest/utils_test/qemu.py]: driver is not defined.

File: virttest/utils_test/qemu.py [hunk 6]
```diff
@@ -142,23 +161,29 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
     :param timeout: Timeout in seconds.
     :return: Session after reboot.
     """
-    verifier_setup_cmd = "verifier.exe /standard /driver %s.sys" % driver
-    if driver:
+    verifier_status, output = check_driver_verifier(session,
+                                                    driver,
+                                                    vm,
```
Comment (pingl2016) [inline virttest/utils_test/qemu.py]: Don't need to pass vm here.

File: virttest/utils_test/qemu.py [hunk 7]
```diff
@@ -142,23 +161,29 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
...
-    output = session.cmd_output(verifier_query_cmd, timeout=timeout)
-    if driver not in output:
-        msg = "Verify device driver failed, "
-        msg += "guest report driver is %s, " % output
-        msg += "expect is '%s'" % driver
-        raise exceptions.TestFail(msg)
+        session = vm.reboot(session)
+        verifier_status, output = check_driver_verifier(session,
+                                                        driver,
+                                                        vm,
```
Comment (pingl2016) [inline virttest/utils_test/qemu.py]: Don't need to pass vm here.

**Comment by pingl2016** [comment]
I found function setup_win_driver_verifier() and clear_win_driver_verifier() come in pairs,  and the former has checked status of driver verifier. So it is unnecessary to check it again in the function clear_win_driver_verifier()

**Comment by suqinhuang** [comment]
some situation like this:

rng_test is sub_test of driver_in_use
clear_win_driver_verifier() is called in rng_test
clear_win_driver_verifier() is called in driver_in_use

clear_win_driver_verifier() will be run twice if didn't check the status

**Comment by pingl2016** [comment]
It's ok for me now.
Acked-by: Ping Li pingl@redhat.com

File: virttest/utils_test/qemu.py [hunk 8]
```diff
@@ -132,6 +132,24 @@ def pin_vm_threads(vm, node):
...
+    Check driver verifier status
+    : param session: VM session
+    : param driver: The driver need to query
+    : param timeout: Timeout in seconds
+    : return: driver verifier status
+    """
+    logging.info("Check %s driver verifier status" % driver)
+    query_cmd = "verifier /querysettings"
+    output = session.cmd_output(query_cmd, timeout=timeout)
+    if driver in output:
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: how about return (driver in output, output)

File: virttest/utils_test/qemu.py [hunk 9]
```diff
@@ -167,11 +184,14 @@ def clear_win_driver_verifier(session, vm, timeout=300):
     :param timeout: Timeout in seconds.
     :return: Session after reboot.
     """
-    verifier_clear_cmd = "verifier.exe /reset"
-    session.cmd(verifier_clear_cmd,
-                timeout=timeout,
-                ignore_all_errors=True)
-    return vm.reboot(session)
+    verifier_status, output = check_driver_verifier(session, driver)
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: where output used?

File: virttest/utils_test/qemu.py [hunk 10]
```diff
@@ -142,20 +160,19 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
...
-        msg += "guest report driver is %s, " % output
-        msg += "expect is '%s'" % driver
-        raise exceptions.TestFail(msg)
+        session = vm.reboot(session)
+        verifier_status, output = check_driver_verifier(session, driver)
+        if not verifier_status:
+            msg = "%s verifier is not enabled, details: %s" % (driver, output)
+            raise exceptions.TestFail(msg)
+    logging.info("%s verifier is enabled already" % driver)
     return session
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: do we really need a session object here?

**Comment by suqinhuang** [comment]
@xutian updated, could you review again

File: virttest/utils_test/qemu.py [hunk 11]
```diff
@@ -140,23 +155,21 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
     :param driver: The driver which needs enable the verifier.
     :param vm: VM object.
     :param timeout: Timeout in seconds.
-    :return: Session after reboot.
     """
-    verifier_setup_cmd = "verifier.exe /standard /driver %s.sys" % driver
-    if driver:
+    verifier_status = check_driver_verifier(session, driver)[1]
+    if not verifier_status:
+        error_context.context("Enable %s driver verifier" % driver)
```
Comment (PyLearner) [inline virttest/utils_test/qemu.py]: hi, if didn't specify the logging function ,will this context be displayed in debug.log,thanks.

File: virttest/utils_test/qemu.py [hunk 12]
```diff
@@ -140,23 +155,22 @@ def setup_win_driver_verifier(session, driver, vm, timeout=300):
...
-        msg += "expect is '%s'" % driver
-        raise exceptions.TestFail(msg)
-    return session
+        session = vm.reboot(session)
+        verifier_status, output = check_driver_verifier(session, driver)
+        if not verifier_status:
+            msg = "%s verifier is not enabled, details: %s" % (driver, output)
+            raise exceptions.TestFail(msg)
+    logging.info("%s verifier is enabled already" % driver)
+    return
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: drop it.

File: virttest/utils_test/qemu.py [hunk 13]
```diff
@@ -165,13 +179,16 @@ def clear_win_driver_verifier(session, vm, timeout=300):
...
-    return vm.reboot(session)
+    verifier_status = check_driver_verifier(session, driver)[1]
+    if verifier_status:
+        logging.info("Clear driver verifier")
+        verifier_clear_cmd = "verifier /reset"
+        session.cmd(verifier_clear_cmd,
+                    timeout=timeout,
+                    ignore_all_errors=True)
+        vm.reboot(session)
+    return
```
Comment (xutian) [inline virttest/utils_test/qemu.py]: same as above.

**Pull request merged**